BigQuery/src/Dataset.php

```diff
@@ -457,4 +381,211 @@ function (array $model) {
+     * }
+     * ```
+     *
+     * @see https://cloud.google.com/bigquery/docs/reference/v2/routines/list List Routines API documentation.
+     *
+     * @param array $options [optional] {
+     *     Configuration options.
+     *
+     *     @type int $maxResults Maximum number of results to return per page.
+     *     @type int $resultLimit Limit the number of results returned in total.
+     *           **Defaults to** `0` (return all results).
+     *     @type string $pageToken A previously-returned page token used to
+     *           resume the loading of results from a specific point.
+     * }
+     * @return ItemIterator<Model>
+     */
+    public function routines(array $options = [])
+    {
+        $resultLimit = $this->pluck('resultLimit', $options, false);
+
+        return new ItemIterator(
+            new PageIterator(
+                function (array $routine) {
+                    return $this->routine($routine['routineReference']['routineId'], $routine);
+                },
+                [$this->connection, 'listRoutines'],
+                $this->identity + $options,
+                [
+                    'itemsKey' => 'routines',
+                    'resultLimit' => $resultLimit
+                ]
+            )
+        );
+    }
+
+    /**
+     * Creates a routine.
+     *
+     * Please note that by default the library will not attempt to retry this
+     * call on your behalf.
+     *
+     * Example:
+     * ```
+     * $routine = $dataset->createRoutine('my_routine', [
+     *     'routineType' => 'SCALAR_FUNCTION',
+     *     'definitionBody' => 'concat(x, "\n", y)',
+     *     'arguments' => [
+     *         [
+     *             'name' => 'x',
+     *             'dataType' => [
+     *                 'typeKind' => 'STRING'
+     *             ]
+     *         ], [
+     *             'name' => 'y',
+     *             'dataType' => [
+     *                 'typeKind' => 'STRING'
+     *             ]
+     *         ]
+     *     ]
+     * ]);
+     * ```
+     *
+     * @see https://cloud.google.com/bigquery/docs/reference/v2/routines/insert Insert Routines API documentation.
+     *
+     * @param string $id The routine ID.
+     * @param array $metadata The available options for metadata are outlined at the
+     *        [Routine Resource API docs](https://cloud.google.com/bigquery/docs/reference/v2/routines#resource).
+     *        Omit `routineReference` as it is computed and appended by the
+     *        client.
+     * @param array $options [optional] Configuration options.
+     * @return Routine
+     */
+    public function createRoutine($id, array $metadata, array $options = [])
+    {
+        $metadata = [
+            'routineReference' => $this->identity + ['routineId' => $id]
+        ] + $metadata;
+
+        $response = $this->connection->insertRoutine(
+            $this->identity
+            + $metadata
+            + ['retries' => 0]
```

Comment:
It looks like this should be below `$options`  so users can still override the number of retries.